### ISSUE TITLE
authentication

### ISSUE THREAD

**Issue by tobit6795** [klaasnicolaas/python-net2grid#81]
Hi,

probably not the right place to ask this question, but i played around this evening with a device called "iona" from EON Group Germany. After a little bit of research i found this repository and for me, it looked like the python lib might work on this device:

`Traceback (most recent call last):
  File "/home/tokem/.local/lib/python3.9/site-packages/net2grid/net2grid.py", line 72, in request
    response.raise_for_status()
  File "/home/tokem/.local/lib/python3.9/site-packages/aiohttp/client_reqrep.py", line 1004, in raise_for_status
    raise ClientResponseError(
aiohttp.client_exceptions.ClientResponseError: 401, message='Unauthorized', url=URL('http://10.10.0.150/meter/now')
`
When i open the device in browser im getting a blank page, but for me it looks like /meter/now is available.
Do you have any idea/information what i could do next? I was successfull  at getting a bearer token +  some JSON data on their public cloud to which the german app is connected - but im more interested in doing stuff like this locally - maybe to integrate it in homeassistant then

greetings

**Comment by klaasnicolaas** [comment]
Currently there is only support for the SmartBridge SBWF3102, which is only sold by Pure Energie under their own white label brand. For any other kind of smartbridge model from Net2Grid I will first have to find out and test what the possibilities are.

**Comment by tobit6795** [comment]
looks like the iona box or a similar device is not listed there. based on the information i got it is a lora gateway.
```
{
    "status": "ok",
    "data": {
        "current_status": "CONNECTED_WITH_METER",
        "meter_txrssi": -62,
        "wlan_rssi": -55,
        "lora_mode": 0,
        "ready_for_meter_inclusion": {
            "inclusion_started": true,
            "KMS_activation_counter": 9,
            "timestamp": "2022-02-09T05:11:32+01:00"
        },
        "contact_with_meter": {
            "connection": true,
            "activation_counter": 9,
            "timestamp": "2020-10-24T10:09:10+02:00"
        },
        "connected_with_meter": {
            "inclusion": true,
            "status": 0,
            "timestamp": "2022-02-09T15:11:32+01:00"
        },
        "timestamp": "2022-02-09T15:14:02+01:00"
    }
}

```

let me now if i can help you somehow

**Comment by klaasnicolaas** [comment]
The question is a bit which API endpoints does it have that we can access. Isn't there something like a model number on the outside?

**Comment by klaasnicolaas** [comment]
You can contact me on Discord (Mister Nicolaz#7558), then we can explore the possibilities further.

**Comment by klaasnicolaas** [comment]
Would still be nice if we can provide support for the iona device.

**Comment by Exquitas** [comment]
Just adding to this discussion as I want to get this working as many other too.

I've found this endpoint in the n2g api docs:
https://api.net2grid.com/reference/v2getlantoken
Which you can call with the Bearer token you receive from the login to their customer frontend (eon smartcontrol etc).

It returns a user_lan_token which expires after 24h.

What I'm struggling with, as I seem unable to catch the calls going to and from the SPLR4911, as my router does not allow me at this point to send everything through a debug proxy and I have no device available right now to intercept the traffic from it, and so I'm not able to figure out how this token is supposed to be used for getting access to /meter/now

So far I've attempted using the mac address as user and the token as password - nope
Same applies for using it as bearer token or as query string.

Potentially this might be the missing key to regain access to the endpoint.

I also wrote an email to net2grid a couple of days ago, but I sadly don't expect an answer from them to be honest ...

**Comment by klaasnicolaas** [comment]
Note that `/meter/now` is not mentioned in those API docs. Unfortunately, I don't have an EON device, so I have to rely on others to test remotely.

**Comment by Exquitas** [comment]
Yes correct, the endpoint on the local device, be it the plug or the dongle without the plug, is not directly related to the enduser interface they offer (to which their API docs are related to).

Basically what we do as endusers (without the local API) is connect to a 'broker' of n2g for one of their clients, which is whitelabeled.
The backend of this entire system is ... N2G.

Same as the hardware, (https://www.net2grid.com/hardware) it's from n2g.
All in all the whole thing is an entire ecosystem running on the same infrastructure. It does not matter if my device came from EON or not as the device technical aspect is the same as the one from any other. The device is sold as white-label product from N2G.

For instance, I can use the N2G api docs to interact with my EON (they call it SmartControl) account, through the Bearer token. So all api calls in their docs are available to me.
The local device however, is sadly nowhere documented.

Weirdly enough, approx 2 months ago they introduced the `user_lan_token` which is approximately around the time (or a bit later) when the firmware rolled out which requires auth for `/meter/now`

PS. EON told me they have no idea and I should possibly try to get in touch with N2G.

**Comment by Exquitas** [comment]
Just an update to this case, as I've received a reply from N2G directly:

<img width="694" alt="image" src="https://github.com/klaasnicolaas/python-gridnet/assets/31363553/e49889f6-27a4-48b6-9e06-72a38caab002">

So in order to access the local API you have to follow these steps:

1. Login to the interface/web app your Energy provider gave you access to. In my case that's https://smartcontrol.eon.de/
2. Grab the bearer token from any API call that has been done in the 'Networking' debug console and save it as you require it as auth for the next call
3. Check the base URL which is used for the actual API call too in your browser networking. In my case that is api.n2g-iona.net
4. Next, make a call to `https://insertYourBackendApiUrlHere/v2/lan/token` with the bearer token from 2.
5. This'll return the auth token for the local api
6. You can now call all endpoints in the local device by setting this header: `Authorization: N2G-LOCAL-LAN token="yourTokenHere"`

Now to the issues with this:
* The Bearer token can *only* be received through logging in in their broker web app. You cannot get the token programmatically through an API call as you as an 'End user' lack the client_id and secret_id which you require to make this call. As per N2G, they intend to implement an End user facing API at some point, but ultimately it is up to your Energy provider (who manages the access to the broker platform) if they make this available to you or not.
* The Bearer token expires in 5 minutes. So you can't really store it for long and reuse it all the time.
* The Local API token expires in 24 hours. So you're forced to call at least once a day to get a new token.

BUT, apparently you can access the local api just fine if your energy provider deactivates the remote access (and management) to the device. But again, this is up to your Energy provider if they do that, can do that, or are even aware of this functionality. I apparently have this functionality through the EON interface, but it does not work. Even by turning on the 'offline mode' the local device still requires auth in my case.

TL;dr: For automating it, it's a mess (I still managed to do it with a lot of extra steps, but nothing I would ship to a userbase to be honest). It's up to your energy provider to re-activate local access as they apparently manage this functionality from their own backend.

**Comment by tobit6795** [comment]
@Exquitas check out https://github.com/tobit6795/homeassistant-smartcontrol
its not pefect, but mostly does it job. it is new to me, that the bearer token expires in 5min, as far as i remember it expires in 60min. Getting the token wasnt a big deal. You can ping me if you want on discord and i can explain it. Anyway thank you for your findings, i just talked with nicholas about this lan token last week. that helps further investigations. im tobit#0746 on dc

**Comment by klaasnicolaas** [comment]
That an Energy provider has the choice to deactivate remote access is known to me, have good contacts with someone at Pure Energie and they chose to deactivate it with a new firmware update.

**Comment by tobit6795** [comment]
````
/info

{
 "id": "84df:0c00:0080:0110",
 "mf": "NET2GRID",
 "model": "iONA4901",
 "fw": "1.7.28",
 "hw": 1,
 "batch": "ION-KES-180802"
}
```

Probably define device lists / dicts?

**Comment by Exquitas** [comment]
@klaasnicolaas 

From what I understood from the Emails between N2G and me (the above posted one and the subsequent emails) is that the Firmware is rolled out by N2G themselves.

The energy companies only give you access to a prettifyed front end with their brand labelling on it, while the entire communication of the physical device that you have, happens directly with N2G and not with the energy company.

The energy company may choose to add some additional things to this front end, but ultimately the entire data from your physical device is sent to N2G.

> that you have to request a token remotely first and then request your data locally

That's correct. You need the Bearer Token first (or access token) which you'll get by logging into their webapp with email + password. Then using this Bearer Token, you can talk to the backend api directly (ie. directly to n2g) and request the user_lan_token , which in turn you have to use to access the local physical device that you have.

I know, it's messy and I don't like it either, but that's how things with N2G devices work currently.

**Comment by klaasnicolaas** [comment]
That's right, N2G is rolling out that firmware. But that was at Pure Energie's request and also only for their customers who have a pure energie meter (PEM). Thanks to the output of @tobit6795 I finally see that there are differences in devices, because Pure Energie users have a different `model` and `batch` value.
